### PR TITLE
[perf] Better CLJC memoization; bounded strategy, fast versions

### PR DESCRIPTION
**Pull request by bshepherdson**
### Description

Expands and refactors `metabase.util.memoize` to support better cross-platform memoization.

**Better basic memoization**

Use `clojure.core.memoize/memo` (and its clone in CLJS) rather than `clojure.core/memoize` for
the "everything forever" strategy.

**Bounded memoization**

There are a few places (eg. `u/kebab-case-en`) where we want zero-overhead *hits* and expect the input space
to be fixed. To guard against runaway memory usage, this **bounded** strategy dumps the entire cache when it
overflows. It logs an INFO level note when that happens - the intent here is that the threshold should never
actually get hit!

**Fast JVM versions**

In the special case where the function to memoize takes exactly 1 argument which is a valid map key, we can
use `ConcurrentHashMap.computeIfAbsent` as the basis of the cache and have much less overhead.

`fast-memo` and `fast-bounded` implement this in CLJ; in CLJS they just call `memo` and `bounded`.

### How to verify

This should be a transparent change that reduces the CPU and memory overhead of memoized functions like
`u/kebab-case-en`. Behavior of the app should not change.

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

**Comment by bshepherdson** [comment]
* **#46199** <a href="https://app.graphite.dev/github/pr/metabase/metabase/46199?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @bshepherdson and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: src/metabase/util/memoize.cljc [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  the input set is larger than expected.
+
+  This is not quite a drop-in replacement for [[memo]] (or `clojure.core/memoize`) because those will *never* call the
+  inner function twice for the same arguments, but this will.
+
+  Default `:bounded/threshold` is 1024."
+  ([f] (bounded f :bounded/threshold 1024))
+  ([f tkey threshold]
+   (assert (= tkey :bounded/threshold) (str "wrong parameter tkey " tkey))
+   (memoize/memoizer f (bounded/bounded-cache-factory {} threshold) {})))
```
Comment (lbrdnk) [inline src/metabase/util/memoize.cljc:40]: Why do we need `tkey` as a parameter?

**Review by lbrdnk — COMMENTED** [review]
LGTM, not approving yet to give a chance @metamben to see the change prior to merging.

File: src/metabase/util/memoize.cljc (hunk 1 above)
Comment (bshepherdson) [reply src/metabase/util/memoize.cljc:40]: We don't, really, but this is how the `core.memoize` library works: `(clojure.core.memoize/lru function-to-memoize :lru/threshold 50)`. Just following along with the pattern.

(These functions pre-date the `[& {:keys [a b]}]` kwargs support I think.)

File: src/metabase/util/memoize.cljc [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  hits very fast. The cost is that it will occasionally recompute even the most frequently hit values.
+
+  The threshold is intended to be big enough to hold everything forever! It's just there as a safety valve in case
+  the input set is larger than expected.
+
+  This is not quite a drop-in replacement for [[memo]] (or `clojure.core/memoize`) because those will *never* call the
+  inner function twice for the same arguments, but this will.
+
+  Default `:bounded/threshold` is 1024."
+  ([f] (bounded f :bounded/threshold 1024))
```
Comment (metamben) [inline src/metabase/util/memoize.cljc:37]: What about making 10000 the default?

File: src/metabase/util/memoize.cljc (hunk 1 above)
Comment (metamben) [reply src/metabase/util/memoize.cljc:40]: They do seem to be using kwargs.

File: src/metabase/util/memoize.cljc [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  ([f tkey threshold]
+   (assert (= tkey :bounded/threshold) (str "wrong parameter tkey " tkey))
+   (memoize/memoizer f (bounded/bounded-cache-factory {} threshold) {})))
+
+(defn fast-memo
+  "Variant of [[memo]] that optimizes a common special case: a function with only a single argument.
```
Comment (metamben) [inline src/metabase/util/memoize.cljc:43]: Would it make sense to point out that types of the keys should have a decent Java `hashCode` method?

File: src/metabase/util/memoize/impl/bounded.cljc [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  (lookup [_ item] (get cache item))
+  (lookup [_ item not-found] (get cache item not-found))
+  (has?   [_ item] (contains? cache item))
+  (hit    [this _item] this)
+  (miss   [_ item result]
+    (if (< (count cache) threshold)
+      (BoundedCache. (assoc cache item result) threshold)
+      ;; It's too big! Throw away the original cache and start over!
+      (do
+        (log/infof "BoundedCache threshold (%d) exceeded - runaway cache? threshold too small?" threshold)
```
Comment (metamben) [inline src/metabase/util/memoize/impl/bounded.cljc:18]: Doesn't this deserve to be a warning? We don't expect this to happen.

File: test/metabase/util/memoize_test.cljc [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+(defn- round-robin [times keyspace f exp-fn]
+  (let [exp-map (into {} (map (juxt identity exp-fn)) keyspace)]
+    (dotimes [_ times]
+      (doseq [k keyspace]
+        (is (= (get exp-map k ::not-found)
+               (f k)))))))
+
+(defn- never-evicts [memoizer]
+  (let [keyspace                 (vec (set (gen/sample gen/string 100)))
```
Comment (metamben) [inline test/metabase/util/memoize_test.cljc:33]: Why not use `gen/set`?

File: test/metabase/util/memoize_test.cljc [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+  (never-evicts #(memo/bounded % :bounded/threshold 100)))
+
+(deftest ^:parallel fast-bounded-never-evicts-if-large-enough-test
+  (never-evicts #(memo/fast-bounded % :bounded/threshold 100)))
+
+(deftest ^:parallel bounded-evicts-when-keyspace-overflows-test
+  (let [n                              100
+        keyspace                       (->> (gen/sample gen/string (* 2 n))
+                                            set
+                                            (take n))
```
Comment (metamben) [inline test/metabase/util/memoize_test.cljc:59]: Why not use `gen/set`?

**Review by metamben — APPROVED** [review]
LGTM, only nits.

File: src/metabase/util/memoize.cljc (hunk 2 above)
Comment (bshepherdson) [reply src/metabase/util/memoize.cljc:37]: This really depends on the size of the inputs and outputs. For `u/kebab-case-en` and friends I know they're small, so it's alright to have 10k. I think 1024 is a reasonable default.

File: src/metabase/util/memoize/impl/bounded.cljc (hunk 4 above)
Comment (bshepherdson) [reply src/metabase/util/memoize/impl/bounded.cljc:18]: I suppose so. It seems a bit trivial for that level, but we would want to see it reported by any user who was seeing it in the wild.

File: test/metabase/util/memoize_test.cljc (hunk 5 above)
Comment (bshepherdson) [reply test/metabase/util/memoize_test.cljc:33]: I didn't know you could pass it the size.

**Pull request merged**